design-proposals/volume-claim-templates.md

```diff
@@ -0,0 +1,181 @@
+
+## Goals
+- Provide a method for creating VMs with PVCs that suits best the user's storage while using CDI volume populators.
+
+## Non Goals
+
+## User Stories
+* As a user creating VMs, I want to be able to create a PVC within the VM spec while getting the best PVC spec that is also compatible with backup and restore solutions.
+
+## Repos
+* **CDI**: Provide API to that receives a PVC spec and in case of missing properties such as VolumeMode and AccessModes fills the optimal values according to the storage class.
+* **CDI**: Update DataImportCron API to create/update VolumeCloneSources that can be consumed as sources for PVC handled by CDI volume populators. VolumeCloneSource will need to be able to be updated with different PVC source. Need to wait for [k8s cross namespace support](https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/) and update clone populator accordingly to allow clone from golden images namespace.
+* **Kubevirt**: Provide API to create a VM with minimal volume information and receive a VM with best suited populated PVC. Adjust InstanceTypes and preferences API to correctly create VM.
+* **Kubevirt**: (optional) Provide virtctl command to generate and create PVCs.
+* **Kubevirt**: Extend virtctl VM create command to use the new API.
+* **SSP**: Manage setting up VolumeCloneSources for commonly supported operating systems updated by DataImportCrons.
+
+*>Note:* Any repo or component that creates VMs (UI, virtctl, tekton pipelines, etc) may want to adjust and prefer using this new API in the VM spec. It should be noted that this can be done incrementally since we do not propose to eliminate the dataVolumeTemplates section right away.
+
+# Design
+
+A new field will be added to the VM object called **VolumeClaimTemplates**. This field will allow the user a way to define and create a PVC owned by the VM while providing partial/full PVC spec embedded in the VM spec. In case of providing full PVC spec with source independent of CDI, no CDI dependency will be required and this operation will not require CDI in the cluster. In case of partial spec we will use CDI API library to fill out these possible fields: VolumeMode, AccessModes, Size(in case of clone) and virtualization optimal storage class (planned work - [design doc PR](https://github.com/kubevirt/community/pull/233)).  
+As described a new API will be created in the CDI API library. This API will receive a partial PVC template and will return a complete PVC spec. VolumeMode and AccessModes will be filled according to storageProfile matching the PVC storageClass. In case of empty storageClass it will rely on either the default StorageClass or if defined `virtualization optimal storage class`(WIP as mentioned). If the of the source of the PVC is VolumeCloneSource Size is also optional and can be filled automatically.  
```

Comment:
A key problem to me is the invisible switching between using a full template and relying on CDI.

It's good for the user, but then I'd add the requirement that ther eis full coverage for all error cases for partial specs, and all of these error cases must be clearly stated as VM error conditions - or through some other means.
IOW: A user should undertstand why something failed. In this case why the completion of a partial spec failed.